lib/render/SliceRenderer.cpp

```diff
@@ -178,144 +188,231 @@ int SliceRenderer::_resetBoxCache()
+        orderedTwoDPoints.pop();
+    }
+    return rectangle2D;
+}
 
-            index += 2;
-            coords[X] += deltas[X];
+// Map our rectangle's 2D edges back into 3D space, to get an
+// ordered list of vertices for our data-enclosing polygon.
+std::vector<glm::vec3> SliceRenderer::_makePolygon3D(const std::vector<_vertex> &vertices, stack<glm::vec2> &polygon2D) const
+{
+    std::vector<glm::vec3> polygon3D;
+    while (!polygon2D.empty()) {
+        glm::vec2 twoDPoint = polygon2D.top();
+        for (auto &vertex : vertices) {
+            if (twoDPoint.x == vertex.twoD.x && twoDPoint.y == vertex.twoD.y) {
+                polygon3D.push_back(glm::vec3(vertex.threeD.x, vertex.threeD.y, vertex.threeD.z));
+                polygon2D.pop();
+                break;
+            }
         }
-        coords[Y] += deltas[Y];
     }
+    return polygon3D;
 }
 
-void SliceRenderer::_populateDataXZ(float *dataValues, Grid *grid) const
+std::vector<glm::vec3> SliceRenderer::_makeRectangle3D(const std::vector<_vertex> &vertices, stack<glm::vec2> &polygon2D) const
 {
-    std::vector<double> deltas = _calculateDeltas();
-    float               varValue, missingValue;
-    std::array<double, 3> coords = {0.0, 0.0, 0.0};
-    coords[X] = _cacheParams.domainMin[X];
-    coords[Y] = _cacheParams.boxMin[Y];
-    coords[Z] = _cacheParams.domainMin[Z];
-
-    int index = 0;
-    for (int j = 0; j < _textureSideSize; j++) {
-        coords[X] = _cacheParams.domainMin[X];
-
-        for (int i = 0; i < _textureSideSize; i++) {
-            varValue = grid->GetValue(coords);
-            missingValue = grid->GetMissingValue();
-            if (varValue == missingValue)
-                dataValues[index + 1] = 1.f;
-            else
-                dataValues[index + 1] = 0.f;
-
-            dataValues[index] = varValue;
+    // Define a rectangle that encloses our polygon in 3D space.  We will sample along this
+    // rectangle to generate our 2D texture.
+    std::vector<glm::vec3> rectangle3D = {glm::vec3(), glm::vec3(), glm::vec3(), glm::vec3()};
+    auto                   inverseProjection = [&](float x, float y) {
+        glm::vec3 point;
+        point = _origin + x * _axis1 + y * _axis2;
+        return point;
+    };
+    rectangle3D = {glm::vec3(), glm::vec3(), glm::vec3(), glm::vec3()};
+    rectangle3D[3] = inverseProjection(_rectangle2D[0].x, _rectangle2D[0].y);
+    rectangle3D[0] = inverseProjection(_rectangle2D[1].x, _rectangle2D[0].y);
+    rectangle3D[1] = inverseProjection(_rectangle2D[1].x, _rectangle2D[1].y);
+    rectangle3D[2] = inverseProjection(_rectangle2D[0].x, _rectangle2D[1].y);
+
+    return rectangle3D;
+}
 
-            index += 2;
-            coords[X] += deltas[X];
-        }
-        coords[Z] += deltas[Z];
-    }
+// Huges-Moller algorithm to get an orthogonal vector
+// https://blog.selfshadow.com/2011/10/17/perp-vectors/
+glm::vec3 SliceRenderer::_getOrthogonal(const glm::vec3 u) const
```

Comment:
While this may be correct, it requires squinting or a proof to confirm. A more self-evident algorithm is to take the cross product with any other vector in the same space

Reply:
From what I've read, there are several approaches to this problem.  Huges-Moller seemed simplest.  If we take a cross product of our normal with any other vector, we could end up with the wrong result.  For example, the normal x normal would be 0. 

Note: I am using your suggested approach for defining _axis2 of the basis function.  From my research, the first axis requires something more along these lines.

Reply:
Thats fine with me, my way also has its drawbacks. Just a note though: normal x normal is not any other vector, it is the same vector, which is why it doesn't work.